File: sudoku.py
#!/usr/bin/env python

from fileinput import input

class Sudoku:

    def getPossible(self, pos):
        possible = {1:1,2:1,3:1,4:1,5:1,6:1,7:1,8:1,9:1}
        grid_corner = self.sector[pos]
        row = pos / self.side
        col = pos % self.side

        # Check 3x3 grid
        for x in [0,1,2,9,10,11,18,19,20]:
            if self.grid[grid_corner+x] in possible: possible.pop(self.grid[grid_corner+x])

        for i in range(self.side):
            if self.grid[row*self.side+i] in possible: possible.pop(self.grid[row*self.side+i])
            if self.grid[i*self.side+col] in possible: possible.pop(self.grid[i*self.side+col])

        return possible

    def solve(self):
        min        = 0
        min_pos    = 0
        options    = 0
        zerocount  = 0
        deductions = 0

        while True:
            min = self.side + 1
            min_pos = -1
            zerocount = 0
            deductions = 0
            for i in range(self.side * self.side):
                if self.grid[i] != 0:
                    continue
                zerocount += 1
                possible = self.getPossible(i)
                count = len(possible)
                if count == 0:
                    return False

                if count == 1:
                    self.grid[i] = possible.keys()[0]
                    deductions += 1
                else:
                    if count < min:
                        min_pos = i
                        min     = count

            if deductions == 0:
                break

        if zerocount == 0:
            return True

        original_grid = self.grid[:]
        possible      = self.getPossible(min_pos)

        for key in possible:
            self.grid[min_pos] = key
            if self.solve(): return True
            self.grid = original_grid[:]
        return False

    def __init__(self):
        self.sector = [  0, 0, 0, 3, 3, 3, 6, 6, 6,
                         0, 0, 0, 3, 3, 3, 6, 6, 6,
                         0, 0, 0, 3, 3, 3, 6, 6, 6,
                        27,27,27,30,30,30,33,33,33,
                        27,27,27,30,30,30,33,33,33,
                        27,27,27,30,30,30,33,33,33,
                        54,54,54,57,57,57,60,60,60,
                        54,54,54,57,57,57,60,60,60,
                        54,54,54,57,57,57,60,60,60 ]
        self.side = 9
        self.grid = [] 

        for line in input():
            self.grid += [int(i) for i in line.split()]

        self.solve()

        for (i, val) in enumerate(self.grid):
            print val,
            if i%self.side == self.side-1:
                print
 
if __name__ == '__main__':
    Sudoku()

File: sudoku.java
import java.util.*;

class sudoku {

    public static final int   side = 9;
    int[] grid;

    int[] getPossible(int pos) {

        int[] possible = {0,1,2,3,4,5,6,7,8,9};

        int row         = pos / side;
        int col         = pos % side;
        int grid_corner = (row / 3) * 27 + (col / 3) * 3;

        // Check 3x3 grid
        possible[ grid[grid_corner   ] ] = 0;
        possible[ grid[grid_corner+1 ] ] = 0;
        possible[ grid[grid_corner+2 ] ] = 0;
        possible[ grid[grid_corner+9 ] ] = 0;
        possible[ grid[grid_corner+10] ] = 0;
        possible[ grid[grid_corner+11] ] = 0;
        possible[ grid[grid_corner+18] ] = 0;
        possible[ grid[grid_corner+19] ] = 0;
        possible[ grid[grid_corner+20] ] = 0;
        
        // Check row
        for (int i = 0; i < side; i++)
            possible[ grid[row*side+i] ] = 0;

        // Check column
        for (int i = 0; i < side; i++)
            possible[ grid[i*side+col] ] = 0;

        return possible;
    }

    int countPossible(int pos) {
        int[] possible = getPossible(pos);
        int tally = 0;
        for (int i = 1; i <= side; i++)
            if (possible[i] != 0)
                tally++;
        return tally;
    }

    int getValue(int pos) {
        int[] possible = getPossible(pos);
        for (int i = 1; i <= side; i++)
            if (possible[i] != 0)
                return possible[i];
        return 0;
    }

    boolean solve() {
        int min         = 0;
        int min_pos     = 0;
        int deductions  = 0;
        int options     = 0;
        int zerocount   = 0;

        while (true) {
            min        = side+1;
            min_pos    = -1;
            deductions = 0;
            zerocount  = 0;
            for (int i = 0; i < side*side; i++) {
                if (grid[i] == 0) {
                    zerocount++;
                    options = countPossible(i);
                    if (options == 0)
                        return false;

                    if (options == 1) {
                        grid[i] = getValue(i);
                        deductions++;
                    }
                    else {
                        if (options <= min) {
                            min_pos = i;
                            min     = options;
                        }
                    }
                }
            }

            if (deductions == 0)
                break;
        }

        if (zerocount == 0)
            return true;

        int[] original_grid   = new int[side*side];
        int[] possible        = getPossible(min_pos);

        for (int i = 0; i < side*side; i++)
            original_grid[i] = grid[i];

        for (int i = 1; i <= side; i++) {
            if (possible[i] != 0) {
                grid[min_pos] = i;
                if (solve())
                    return true;
                else {
                    for (int j = 0; j < side*side; j++)
                        grid[j] = original_grid[j];
                }
            }
        }
        return false;
    }

    sudoku() {
        grid = new int[side * side];
        Scanner scan = new Scanner(System.in);

        for (int i = 0; i < side * side; i++)
            grid[i] = scan.nextInt();

        solve();

        for (int i = 0; i < side; i++) {
            for (int j = 0; j < side; j++)
                System.out.print(" "+grid[i*side+j]);
            System.out.println("");
        }
    }

    public static void main(String[] args) {
        long start = System.currentTimeMillis();
        new sudoku();
        System.out.println("Execution time: "+(System.currentTimeMillis()-start)+" millisec.");
    }
}
